ls1intum/Artemis
# Feature/intellij basic plugin

### Checklist

### Motivation and Context


This is the first version of the ArTEMiS plugin called OrION (https://github.com/ls1intum/Orion). In order for the IntelliJ plugin to work correctly, it has to be able to communicate with the Angular webapp. The bridge between the Java and the Typescript code has to provide the following functionalities:

- Git clone, push, add, commit calls from TS to Java
- Notifications about the status of the IDE (IntelliJ) from Java to TS
- Some sort of check, in order to find out whether the browser opening the webapp belongs to an IDE (IntelliJ)

**As previously discussed with @thilo-behnke  I will write the tests for this PR after the first draft has been reviewed, since there might be some bigger changes depending on the review.**

### Description

I've added a new JavaBridgeService which provides the most important functionalities described above. This services can communicate with the IDE by calling functions directly on the window object of the browser. 
Downcalls (from TS to Java) are executed by calling the in the intellij interface specified methods. The actual JS object in the window is set by the plugin in Intellij itself. 
Upcalls (from Java to TS) are executed by calling the methods of the JavaBridgeService, which has to be set at the start of the app, so that the IDE can immediately communicate with the webapp as soon as the page has been loaded. 

### Steps for Testing


1. Install the IntelliJ plugin (https://github.com/ls1intum/Orion/releases/tag/0.1.1-alpha)
Installation:
![orion_installation](https://user-images.githubusercontent.com/27979674/64041207-39966500-cb5f-11e9-9ee5-2541d301a49c.gif)

Setup:
![orion_setup](https://user-images.githubusercontent.com/27979674/64041226-41eea000-cb5f-11e9-961f-6f019ed89c37.gif)

**The settings button here is only temporary and will be removed in future versions!**

2. Open the ArTEMiS tool window on the right side in IntelliJ
3. Open any course in ArTEMiS with programming exercises.
4. Click on "Open in IntelliJ" in order to clone and open the exercise
5. Change anything in the source files of the opened project
6. Click on "Submit" in order to add all changes, commit them and push them to master
7. All changes should be committed and visible in Bitbucket

### Screenshots


![orion_first_workflow](https://user-images.githubusercontent.com/27979674/63797373-41100100-c908-11e9-8c4c-f85c9dee0968.gif)
